File: go/block.go
package embedkv

import "encoding/binary"

var le = binary.LittleEndian

// StorageHeader is the 28-byte header stored at the start of block 0.
// The block's CRC32 is stored separately in the last 4 bytes of the block.
type StorageHeader struct {
	BlockType    uint8
	Magic        [3]byte
	VersionMajor uint16
	VersionMinor uint16
	BlockSize    uint32
	BlockCount   uint32
	ReplicaID    uint32
	FormatSeq    uint32
	Flags        uint32
}

func marshalStorageHeader(h *StorageHeader, buf []byte) {
	buf[0] = h.BlockType
	buf[1] = h.Magic[0]
	buf[2] = h.Magic[1]
	buf[3] = h.Magic[2]
	le.PutUint16(buf[4:], h.VersionMajor)
	le.PutUint16(buf[6:], h.VersionMinor)
	le.PutUint32(buf[8:], h.BlockSize)
	le.PutUint32(buf[12:], h.BlockCount)
	le.PutUint32(buf[16:], h.ReplicaID)
	le.PutUint32(buf[20:], h.FormatSeq)
	le.PutUint32(buf[24:], h.Flags)
}

// unmarshalStorageHeader deserializes the header and verifies the block's
// trailing CRC32. Returns true only when the CRC is valid.
func unmarshalStorageHeader(buf []byte, h *StorageHeader) bool {
	h.BlockType = buf[0]
	h.Magic[0] = buf[1]
	h.Magic[1] = buf[2]
	h.Magic[2] = buf[3]
	h.VersionMajor = le.Uint16(buf[4:])
	h.VersionMinor = le.Uint16(buf[6:])
	h.BlockSize = le.Uint32(buf[8:])
	h.BlockCount = le.Uint32(buf[12:])
	h.ReplicaID = le.Uint32(buf[16:])
	h.FormatSeq = le.Uint32(buf[20:])
	h.Flags = le.Uint32(buf[24:])
	return verifyBlockCRC(buf)
}

// RecordDescriptorHeader is the fixed 32-byte header of a record descriptor block.
// Key bytes follow immediately at offset 32; value payload starts at offset 32+KeySize.
// The block's CRC32 is stored separately in the last 4 bytes of the block.
type RecordDescriptorHeader struct {
	BlockType        uint8
	HeaderSize       uint8
	KeySize          uint16 // key length in bytes (UTF-8)
	Generation       uint32
	TotalSize        uint32
	FirstPayloadSize uint32
	ChunkCount       uint32
	NextChunk        uint32
	Flags            uint32 // internal flags, reserved — always 0
	UserFlags        uint32 // user-defined flags; preserved verbatim
}

func marshalDescriptorHeader(h *RecordDescriptorHeader, buf []byte) {
	buf[0] = h.BlockType
	buf[1] = h.HeaderSize
	le.PutUint16(buf[2:], h.KeySize)
	le.PutUint32(buf[4:], h.Generation)
	le.PutUint32(buf[8:], h.TotalSize)
	le.PutUint32(buf[12:], h.FirstPayloadSize)
	le.PutUint32(buf[16:], h.ChunkCount)
	le.PutUint32(buf[20:], h.NextChunk)
	le.PutUint32(buf[24:], h.Flags)
	le.PutUint32(buf[28:], h.UserFlags)
}

// unmarshalDescriptorHeader deserializes the header and verifies the block's
// trailing CRC32. Returns true only when the CRC is valid.
func unmarshalDescriptorHeader(buf []byte, h *RecordDescriptorHeader) bool {
	h.BlockType = buf[0]
	h.HeaderSize = buf[1]
	h.KeySize = le.Uint16(buf[2:])
	h.Generation = le.Uint32(buf[4:])
	h.TotalSize = le.Uint32(buf[8:])
	h.FirstPayloadSize = le.Uint32(buf[12:])
	h.ChunkCount = le.Uint32(buf[16:])
	h.NextChunk = le.Uint32(buf[20:])
	h.Flags = le.Uint32(buf[24:])
	h.UserFlags = le.Uint32(buf[28:])
	return verifyBlockCRC(buf)
}

// ValueChunkHeader is the fixed 20-byte header of a value chunk block.
// The block's CRC32 is stored separately in the last 4 bytes of the block.
type ValueChunkHeader struct {
	BlockType       uint8
	HeaderSize      uint8
	Flags           uint16
	OwnerDescriptor uint32
	ChunkIndex      uint32
	PayloadSize     uint32
	NextChunk       uint32
}

func marshalChunkHeader(h *ValueChunkHeader, buf []byte) {
	buf[0] = h.BlockType
	buf[1] = h.HeaderSize
	le.PutUint16(buf[2:], h.Flags)
	le.PutUint32(buf[4:], h.OwnerDescriptor)
	le.PutUint32(buf[8:], h.ChunkIndex)
	le.PutUint32(buf[12:], h.PayloadSize)
	le.PutUint32(buf[16:], h.NextChunk)
}

// unmarshalChunkHeader deserializes the header and verifies the block's
// trailing CRC32. Returns true only when the CRC is valid.
func unmarshalChunkHeader(buf []byte, h *ValueChunkHeader) bool {
	h.BlockType = buf[0]
	h.HeaderSize = buf[1]
	h.Flags = le.Uint16(buf[2:])
	h.OwnerDescriptor = le.Uint32(buf[4:])
	h.ChunkIndex = le.Uint32(buf[8:])
	h.PayloadSize = le.Uint32(buf[12:])
	h.NextChunk = le.Uint32(buf[16:])
	return verifyBlockCRC(buf)
}

// classifyBlock returns whether buf is a free, non-free valid, or garbage block.
// Per ARCH §9: check type+CRC first; if valid → non-free. If first byte is 0x00/0xFF → free.
// Otherwise → garbage (caller decides what to do).
type blockClass int

const (
	blockClassFree    blockClass = iota
	blockClassValid              // valid non-free block with good CRC
	blockClassGarbage            // corrupted or unknown
)

func classifyBlock(buf []byte) blockClass {
	if len(buf) == 0 {
		return blockClassFree
	}
	first := buf[0]
	switch first {
	case BlockTypeStorageHeader:
		var h StorageHeader
		if unmarshalStorageHeader(buf, &h) {
			return blockClassValid
		}
	case BlockTypeRecordDescriptor:
		var h RecordDescriptorHeader
		if uint32(len(buf)) >= RecordDescriptorHeaderSize+BlockCRCSize && unmarshalDescriptorHeader(buf, &h) {
			return blockClassValid
		}
	case BlockTypeValueChunk:
		var h ValueChunkHeader
		if unmarshalChunkHeader(buf, &h) {
			return blockClassValid
		}
	}
	if first == 0x00 || first == 0xFF {
		return blockClassFree
	}
	return blockClassGarbage
}

// isFreeCandidate returns true if the block can be allocated as free space.
func isFreeCandidate(buf []byte) bool {
	return classifyBlock(buf) == blockClassFree
}

// firstPayloadCapacity returns how many value bytes fit in a descriptor block
// after the header, the key, and the trailing block CRC.
func firstPayloadCapacity(blockSize, keySize uint32) uint32 {
	return blockSize - RecordDescriptorHeaderSize - keySize - BlockCRCSize
}

// chunkPayloadCapacity returns how many value bytes fit in a value chunk block
// after the header and the trailing block CRC.
func chunkPayloadCapacity(blockSize uint32) uint32 {
	return blockSize - ValueChunkHeaderSize - BlockCRCSize
}

// makeFreeBuf returns a block-sized buffer filled with the given free pattern (0x00 or 0xFF).
func makeFreeBuf(blockSize uint32, pattern byte) []byte {
	buf := make([]byte, blockSize)
	if pattern != 0x00 {
		for i := range buf {
			buf[i] = pattern
		}
	}
	return buf
}

File: rust/src/block.rs
use crate::crc::verify_block_crc;
use crate::format::*;

// ── StorageHeader ──────────────────────────────────────────────────────────

/// 28-byte storage header stored at the start of block 0.
/// The block's CRC32 is stored separately in the last 4 bytes of the block.
#[derive(Debug, Clone, Default)]
pub struct StorageHeader {
    pub block_type: u8,
    pub magic: [u8; 3],
    pub version_major: u16,
    pub version_minor: u16,
    pub block_size: u32,
    pub block_count: u32,
    pub replica_id: u32,
    pub format_seq: u32,
    pub flags: u32,
}

pub fn marshal_storage_header(h: &StorageHeader, buf: &mut [u8]) {
    buf[0] = h.block_type;
    buf[1] = h.magic[0];
    buf[2] = h.magic[1];
    buf[3] = h.magic[2];
    buf[4..6].copy_from_slice(&h.version_major.to_le_bytes());
    buf[6..8].copy_from_slice(&h.version_minor.to_le_bytes());
    buf[8..12].copy_from_slice(&h.block_size.to_le_bytes());
    buf[12..16].copy_from_slice(&h.block_count.to_le_bytes());
    buf[16..20].copy_from_slice(&h.replica_id.to_le_bytes());
    buf[20..24].copy_from_slice(&h.format_seq.to_le_bytes());
    buf[24..28].copy_from_slice(&h.flags.to_le_bytes());
}

/// Deserializes the storage header and verifies the block's trailing CRC32.
/// Returns `true` only when the CRC is valid.
pub fn unmarshal_storage_header(buf: &[u8], h: &mut StorageHeader) -> bool {
    h.block_type = buf[0];
    h.magic = [buf[1], buf[2], buf[3]];
    h.version_major = u16::from_le_bytes([buf[4], buf[5]]);
    h.version_minor = u16::from_le_bytes([buf[6], buf[7]]);
    h.block_size = u32::from_le_bytes([buf[8], buf[9], buf[10], buf[11]]);
    h.block_count = u32::from_le_bytes([buf[12], buf[13], buf[14], buf[15]]);
    h.replica_id = u32::from_le_bytes([buf[16], buf[17], buf[18], buf[19]]);
    h.format_seq = u32::from_le_bytes([buf[20], buf[21], buf[22], buf[23]]);
    h.flags = u32::from_le_bytes([buf[24], buf[25], buf[26], buf[27]]);
    verify_block_crc(buf)
}

// ── RecordDescriptorHeader ─────────────────────────────────────────────────

/// Fixed 32-byte header of a record descriptor block.
/// Key bytes follow at offset 32; value payload starts at 32 + key_size.
/// The block's CRC32 is stored separately in the last 4 bytes of the block.
#[derive(Debug, Clone, Default)]
pub struct RecordDescriptorHeader {
    pub block_type: u8,
    pub header_size: u8,
    pub key_size: u16,
    pub generation: u32,
    pub total_size: u32,
    pub first_payload_size: u32,
    pub chunk_count: u32,
    pub next_chunk: u32,
    pub flags: u32,      // internal flags, reserved — always 0
    pub user_flags: u32, // user-defined flags; preserved verbatim
}

pub fn marshal_descriptor_header(h: &RecordDescriptorHeader, buf: &mut [u8]) {
    buf[0] = h.block_type;
    buf[1] = h.header_size;
    buf[2..4].copy_from_slice(&h.key_size.to_le_bytes());
    buf[4..8].copy_from_slice(&h.generation.to_le_bytes());
    buf[8..12].copy_from_slice(&h.total_size.to_le_bytes());
    buf[12..16].copy_from_slice(&h.first_payload_size.to_le_bytes());
    buf[16..20].copy_from_slice(&h.chunk_count.to_le_bytes());
    buf[20..24].copy_from_slice(&h.next_chunk.to_le_bytes());
    buf[24..28].copy_from_slice(&h.flags.to_le_bytes());
    buf[28..32].copy_from_slice(&h.user_flags.to_le_bytes());
}

/// Deserializes the descriptor header and verifies the block's trailing CRC32.
/// Returns `true` only when the CRC is valid.
pub fn unmarshal_descriptor_header(buf: &[u8], h: &mut RecordDescriptorHeader) -> bool {
    h.block_type = buf[0];
    h.header_size = buf[1];
    h.key_size = u16::from_le_bytes([buf[2], buf[3]]);
    h.generation = u32::from_le_bytes([buf[4], buf[5], buf[6], buf[7]]);
    h.total_size = u32::from_le_bytes([buf[8], buf[9], buf[10], buf[11]]);
    h.first_payload_size = u32::from_le_bytes([buf[12], buf[13], buf[14], buf[15]]);
    h.chunk_count = u32::from_le_bytes([buf[16], buf[17], buf[18], buf[19]]);
    h.next_chunk = u32::from_le_bytes([buf[20], buf[21], buf[22], buf[23]]);
    h.flags = u32::from_le_bytes([buf[24], buf[25], buf[26], buf[27]]);
    h.user_flags = u32::from_le_bytes([buf[28], buf[29], buf[30], buf[31]]);
    verify_block_crc(buf)
}

// ── ValueChunkHeader ───────────────────────────────────────────────────────

/// Fixed 20-byte header of a value chunk block.
/// The block's CRC32 is stored separately in the last 4 bytes of the block.
#[derive(Debug, Clone, Default)]
pub struct ValueChunkHeader {
    pub block_type: u8,
    pub header_size: u8,
    pub flags: u16,
    pub owner_descriptor: u32,
    pub chunk_index: u32,
    pub payload_size: u32,
    pub next_chunk: u32,
}

pub fn marshal_chunk_header(h: &ValueChunkHeader, buf: &mut [u8]) {
    buf[0] = h.block_type;
    buf[1] = h.header_size;
    buf[2..4].copy_from_slice(&h.flags.to_le_bytes());
    buf[4..8].copy_from_slice(&h.owner_descriptor.to_le_bytes());
    buf[8..12].copy_from_slice(&h.chunk_index.to_le_bytes());
    buf[12..16].copy_from_slice(&h.payload_size.to_le_bytes());
    buf[16..20].copy_from_slice(&h.next_chunk.to_le_bytes());
}

/// Deserializes the chunk header and verifies the block's trailing CRC32.
/// Returns `true` only when the CRC is valid.
pub fn unmarshal_chunk_header(buf: &[u8], h: &mut ValueChunkHeader) -> bool {
    h.block_type = buf[0];
    h.header_size = buf[1];
    h.flags = u16::from_le_bytes([buf[2], buf[3]]);
    h.owner_descriptor = u32::from_le_bytes([buf[4], buf[5], buf[6], buf[7]]);
    h.chunk_index = u32::from_le_bytes([buf[8], buf[9], buf[10], buf[11]]);
    h.payload_size = u32::from_le_bytes([buf[12], buf[13], buf[14], buf[15]]);
    h.next_chunk = u32::from_le_bytes([buf[16], buf[17], buf[18], buf[19]]);
    verify_block_crc(buf)
}

// ── Block classification ───────────────────────────────────────────────────

#[derive(Debug, PartialEq, Eq)]
pub enum BlockClass {
    Free,
    Valid,
    Garbage,
}

pub fn classify_block(buf: &[u8]) -> BlockClass {
    if buf.is_empty() {
        return BlockClass::Free;
    }
    let len = buf.len() as u32;
    match buf[0] {
        BLOCK_TYPE_STORAGE_HEADER if len >= STORAGE_HEADER_SIZE + BLOCK_CRC_SIZE => {
            let mut h = StorageHeader::default();
            if unmarshal_storage_header(buf, &mut h) {
                return BlockClass::Valid;
            }
        }
        BLOCK_TYPE_RECORD_DESCRIPTOR if len >= RECORD_DESCRIPTOR_HEADER_SIZE + BLOCK_CRC_SIZE => {
            let mut h = RecordDescriptorHeader::default();
            if unmarshal_descriptor_header(buf, &mut h) {
                return BlockClass::Valid;
            }
        }
        BLOCK_TYPE_VALUE_CHUNK if len >= VALUE_CHUNK_HEADER_SIZE + BLOCK_CRC_SIZE => {
            let mut h = ValueChunkHeader::default();
            if unmarshal_chunk_header(buf, &mut h) {
                return BlockClass::Valid;
            }
        }
        _ => {}
    }
    if buf[0] == 0x00 || buf[0] == 0xFF {
        BlockClass::Free
    } else {
        BlockClass::Garbage
    }
}

pub fn is_free_candidate(buf: &[u8]) -> bool {
    classify_block(buf) == BlockClass::Free
}

/// Value bytes that fit in a descriptor block after the header, the key, and
/// the trailing block CRC.
pub fn first_payload_capacity(block_size: u32, key_size: u32) -> u32 {
    block_size - RECORD_DESCRIPTOR_HEADER_SIZE - key_size - BLOCK_CRC_SIZE
}

/// Value bytes that fit in a value chunk block after the header and the
/// trailing block CRC.
pub fn chunk_payload_capacity(block_size: u32) -> u32 {
    block_size - VALUE_CHUNK_HEADER_SIZE - BLOCK_CRC_SIZE
}

/// Returns a block-sized buffer filled with `pattern` (0x00 or 0xFF).
pub fn make_free_buf(block_size: u32, pattern: u8) -> Vec<u8> {
    vec![pattern; block_size as usize]
}
